cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -301,80 +694,305 @@ func generateMountsModeRecover(mst *initramfsMountsState) error {
+	// TODO: this prompts again for a recover key, but really this is the
+	// reinstall key we will prompt for
+	// TODO: we should somehow customize the prompt to mention what key we need
+	// the user to enter, and what we are unlocking (as currently the prompt
+	// says "recovery key" and the partition UUID for what is being unlocked)
+	unlockRes, unlockErr := secbootUnlockVolumeUsingSealedKeyIfEncrypted(m.disk, "ubuntu-save", saveFallbackKey, unlockOpts)
+	if err := m.setUnlockStateWithFallbackKey("ubuntu-save", unlockRes, unlockErr); err != nil {
+		return nil, err
+	}
+	if unlockErr != nil {
+		// all done, nothing left to try and mount, everything failed
+		return nil, nil
+	}
+
+	// otherwise we unlocked it, so go mount it
+	return m.mountSave, nil
+}
+
+func (m *stateMachine) mountSave() (stateFunc, error) {
+	saveDev := m.degradedState.partition("ubuntu-save").Device
+	// TODO: should we fsck ubuntu-save ?
+	mountErr := doSystemdMount(saveDev, boot.InitramfsUbuntuSaveDir, nil)
+	if err := m.setMountState("ubuntu-save", boot.InitramfsUbuntuSaveDir, mountErr); err != nil {
+		return nil, err
+	}
+	// all done, nothing left to try and mount
+	return nil, nil
+}
+
+func generateMountsModeRecover(mst *initramfsMountsState) error {
+	// steps 1 and 2 are shared with install mode
+	if err := generateMountsCommonInstallRecover(mst); err != nil {
 		return err
 	}
 
-	// 3.1. mount ubuntu-save (if present)
-	haveSave, err := maybeMountSave(disk, boot.InitramfsHostWritableDir, unlockRes.IsDecryptedDevice, nil)
+	// get the disk that we mounted the ubuntu-seed partition from as a
+	// reference point for future mounts
+	disk, err := disks.DiskFromMountPoint(boot.InitramfsUbuntuSeedDir, nil)
 	if err != nil {
 		return err
 	}
 
-	// 3.2 verify that the host ubuntu-data comes from where we expect it to
-	// right device
-	diskOpts := &disks.Options{}
-	if unlockRes.IsDecryptedDevice {
-		// then we need to specify that the data mountpoint is expected to be a
-		// decrypted device, applies to both ubuntu-data and ubuntu-save
-		diskOpts.IsDecryptedDevice = true
-	}
+	// 3. run the state machine logic for mounting partitions, this involves
+	//    trying to unlock then mount ubuntu-data, and then unlocking and
+	//    mounting ubuntu-save
+	//    see the state* functions for details of what each step does and
+	//    possible transition points
+
+	machine, err := func() (machine *stateMachine, err error) {
+		// ensure that the last thing we do after mounting everything is to lock
+		// access to sealed keys
+		defer func() {
+			if err := secbootLockTPMSealedKeys(); err != nil {
+				logger.Noticef("error locking access to sealed keys: %v", err)
+			}
+		}()
 
-	matches, err = disk.MountPointIsFromDisk(boot.InitramfsHostUbuntuDataDir, diskOpts)
+		// first state to execute is to unlock ubuntu-data with the run key
+		machine = newStateMachine(disk)
+		for {
+			final, err := machine.execute()
```

Comment:
(nitpick^4) we use "finished" in the execute() and final here, same terms would be slightly nicer I feel

Reply:
@pedronis did this one in https://github.com/snapcore/snapd/pull/9619